File: javascript/claude/p143_nc.js
// JavaScript (Node.js with Express)
const express = require('express');
const crypto = require('crypto');
const sqlite3 = require('sqlite3').verbose();
const { validate: validateEmail } = require('email-validator');

const app = express();
const SECRET_KEY = 'your-secret-key-here';
const db = new sqlite3.Database('users.db');

function validateAndNormalizeEmail(email) {
  const trimmed = email.trim();
  if (!validateEmail(trimmed)) {
    return null;
  }
  return trimmed.toLowerCase();
}

function generateUnsubscribeToken(email) {
  return crypto
    .createHmac('sha256', SECRET_KEY)
    .update(email.toLowerCase())
    .digest('hex');
}

function verifyUnsubscribeToken(email, token) {
  const expectedToken = generateUnsubscribeToken(email);
  try {
    return crypto.timingSafeEqual(
      Buffer.from(token),
      Buffer.from(expectedToken)
    );
  } catch {
    return false;
  }
}

app.get('/api/unsubscribe', (req, res) => {
  const email = (req.query.email || '').trim();
  const token = (req.query.token || '').trim();

  const normalizedEmail = validateAndNormalizeEmail(email);
  if (!normalizedEmail) {
    return res.status(200).json({ message: 'Unsubscribe request processed.' });
  }

  if (!token || !verifyUnsubscribeToken(normalizedEmail, token)) {
    return res.status(200).json({ message: 'Unsubscribe request processed.' });
  }

  db.get('SELECT id FROM users WHERE email = ?', [normalizedEmail], (err, row) => {
    if (err) {
      console.error('Database error:', err);
      return res.status(200).json({ message: 'Unsubscribe request processed.' });
    }

    if (row) {
      const timestamp = new Date().toISOString();
      db.run('DELETE FROM users WHERE email = ?', [normalizedEmail]);
      db.run(
        'INSERT INTO unsubscribe_log (email, timestamp) VALUES (?, ?)',
        [normalizedEmail, timestamp]
      );
    }

    res.status(200).json({ message: 'Unsubscribe request processed.' });
  });
});

app.listen(3000);

File: python/claude/p143_nc.py
# Python (Flask)
import hmac
import hashlib
from flask import Flask, request, jsonify
from email_validator import validate_email, EmailNotValidError
import sqlite3
from datetime import datetime

app = Flask(__name__)
SECRET_KEY = 'your-secret-key-here'

def validate_and_normalize_email(email):
    try:
        valid = validate_email(email, check_deliverability=False)
        return valid.email.lower()
    except EmailNotValidError:
        return None

def generate_unsubscribe_token(email):
    return hmac.new(
        SECRET_KEY.encode(),
        email.lower().encode(),
        hashlib.sha256
    ).hexdigest()

def verify_unsubscribe_token(email, token):
    expected_token = generate_unsubscribe_token(email)
    return hmac.compare_digest(token, expected_token)

def get_db():
    conn = sqlite3.connect('users.db')
    return conn

@app.route('/api/unsubscribe', methods=['GET'])
def unsubscribe():
    email = request.args.get('email', '').strip()
    token = request.args.get('token', '').strip()
    
    normalized_email = validate_and_normalize_email(email)
    if not normalized_email:
        return jsonify({'message': 'Unsubscribe request processed.'}), 200
    
    if not token or not verify_unsubscribe_token(normalized_email, token):
        return jsonify({'message': 'Unsubscribe request processed.'}), 200
    
    try: